presto-main/src/test/java/io/prestosql/sql/planner/TestLogicalPlanner.java

```diff
@@ -801,4 +805,71 @@ public void testDistributedSort()
                                                 tableScan("orders", ImmutableMap.of(
                                                         "ORDERKEY", "orderkey")))))));
     }
+
+    @Test
+    public void testRemoveRedundantLimitNode()
+    {
+        assertFalse(
+                searchFrom(plan("SELECT count(*) FROM orders LIMIT 10", LogicalPlanner.Stage.OPTIMIZED).getRoot())
+                        .where(isInstanceOfAny(LimitNode.class))
+                        .matches(),
+                "Unexpected node for the above query");
+
+        assertPlan(
+                "SELECT orderkey, count(*) FROM orders GROUP BY orderkey LIMIT 10",
+                output(
+                        limit(10,
+                                anyTree(
+                                        tableScan("orders")))));
+    }
+
+    @Test
+    public void testRemoveSingleRowSort()
+    {
+        assertFalse(
+                searchFrom(plan("SELECT count(*) FROM orders ORDER BY 1", LogicalPlanner.Stage.OPTIMIZED).getRoot())
+                        .where(isInstanceOfAny(SortNode.class))
+                        .matches(),
+                "Unexpected node for the above query");
+
+        assertPlan(
+                "SELECT orderkey, count(*) FROM orders GROUP BY orderkey ORDER BY 1",
+                anyTree(
+                        node(SortNode.class,
+                                anyTree(
+                                        tableScan("orders")))));
+    }
+
+    public void testReplaceTopNWithSort()
+    {
+        assertFalse(
+                searchFrom(plan("SELECT count(*) FROM orders ORDER BY 1 LIMIT 10", LogicalPlanner.Stage.OPTIMIZED).getRoot())
+                        .where(isInstanceOfAny(TopNNode.class))
+                        .matches(),
+                "Unexpected node for the above query");
+
+        assertPlan(
+                "SELECT orderkey, count(*) FROM orders GROUP BY orderkey ORDER BY 1 LIMIT 10",
+                output(
+                        node(TopNNode.class,
+                                anyTree(
+                                        tableScan("orders")))));
+    }
+
+    @Test
+    public void testRemoveSingleRowDistinctLimit()
+    {
+        assertFalse(
+                searchFrom(plan("SELECT distinct(c) FROM (SELECT count(*) as c FROM orders) LIMIT 10", LogicalPlanner.Stage.OPTIMIZED).getRoot())
+                        .where(isInstanceOfAny(DistinctLimitNode.class))
```

Comment:
Also check there is no `MarkDistinctNode`. You could also extract a method from this and reuse in assertion below, like:
```
assertFalse(planContainsDistinctNode("SELECT distinct(c) FROM (SELECT count(*) as c FROM orders) LIMIT 10");
assertTrue(planContainsDistinctNode("SELECT distinct(c) FROM (SELECT count(*) as c FROM orders GROUP BY orderkey) LIMIT 10"));
```

Please do the same for `TopN` and `Sort`.